File: 01 09 25/swapnodepairs.go
// O(n)

package main

import "fmt"

// Definition for singly-linked list.
type ListNode struct {
    Val  int
    Next *ListNode
}

func swapPairs(head *ListNode) *ListNode {
    dummy := &ListNode{Next: head}
    prev := dummy

    for head != nil && head.Next != nil {
        first := head
        second := head.Next

        // Swap
        prev.Next = second
        first.Next = second.Next
        second.Next = first

        // Move pointers
        prev = first
        head = first.Next
    }

    return dummy.Next
}

func swapPairsRecursive(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
    }

    first := head
    second := head.Next

    // Recursively swap the rest
    first.Next = swapPairs(second.Next)
    second.Next = first

    return second
}

// Helper functions for testing
func buildList(arr []int) *ListNode {
    dummy := &ListNode{}
    curr := dummy
    for _, v := range arr {
        curr.Next = &ListNode{Val: v}
        curr = curr.Next
    }
    return dummy.Next
}

func printList(node *ListNode) {
    for node != nil {
        fmt.Print(node.Val, " ")
        node = node.Next
    }
    fmt.Println()
}

func main() {
    head := buildList([]int{1, 2, 3, 4})
    swapped := swapPairs(head)
    printList(swapped) // Output: 2 1 4 3
}

File: 01 09 25/swapnodepairs.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        # Count the total number of nodes
        def count_nodes(node):
            cnt = 0
            while node:
                cnt += 1
                node = node.next
            return cnt

        total = count_nodes(head)
        dummy = ListNode(0)
        dummy.next = head
        prev = dummy

        while total >= k:
            curr = prev.next
            nxt = curr.next
            # Reverse k nodes
            for _ in range(1, k):
                curr.next = nxt.next
                nxt.next = prev.next
                prev.next = nxt
                nxt = curr.next
            prev = curr
            total -= k

        return dummy.next
